benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,228 @@
+
+@define
+class PathConstructor:
+    """A class to construct the hierarchical path of a result object.
+
+    This class is used to encapsulate the construction of a respective path depending
+    on where the object should be stored. Since different storage backends have
+    different requirements for the path, and some like the used S3 Bucket
+    not even support folders, the path uses its variables to construct the path
+    in the order of the variables with a separator depending on the strategy.
+    For compatibility reasons, the path is sanitized to only contain the following
+    allowed characters:
+    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+    """
+
+    benchmark_name: str = field(validator=instance_of(str))
+    """The name of the benchmark for which the path should be constructed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    latest_baybe_tag: str = field(validator=instance_of(str))
+    """The latest BayBE version tag."""
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    commit_hash: str = field(validator=instance_of(str))
+    """The latest commit hash of the current checkout branch."""
+
+    @branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        return current_branch
+
+    def _string_sanitizer(self, string: str) -> str:
+        """Replace not allowed chars for filesystems in a string."""
+        ALLOWED_CHARS = (
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+        )
+        return "".join([char if char in ALLOWED_CHARS else "-" for char in string])
+
+    def get_path(self, strategy: PathStrategy) -> str:
+        """Construct the path of a result object.
+
+        Creates the path depending on the chosen strategy by concatenating
+        the experiment identifier, the branch, the BayBE-version,
+        the start date without time and the commit hash.
+
+        Args:
+            strategy: The strategy to construct the path.
+
+        Returns:
+            The path to persist the object. Can be a file name or
+            a folder path as a string.
+        """
+        separator = "/" if strategy == PathStrategy.HIERARCHICAL else "_"
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return path
+
+
+class PersistenceHandlingInterface(Protocol):
```

Comment:
To me, the naming seems overly complicated. Why not simply `ObjectWriter`? This is simpler and even more general in that it would in principle also cover non-persistent stuff such as writing to `stdout`

Reply:
I thought maybe we have to load data at some point. Thats why I have chosen the more generic name. But it would be fine for me to change it now and decide how to deal with reading once its there. I will do that for now.

Reply:
I think `ObjectStorage` is good.